File: plugins/sql/mssql/payload.php
<?php

//            ---------------------------------------------------
//                              Omega Framework                                
//            ---------------------------------------------------
//                  Copyright (C) <2020>  <Entynetproject>       
//
//        This program is free software: you can redistribute it and/or modify
//        it under the terms of the GNU General Public License as published by
//        the Free Software Foundation, either version 3 of the License, or
//        any later version.
//
//        This program is distributed in the hope that it will be useful,
//        but WITHOUT ANY WARRANTY; without even the implied warranty of
//        MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
//        GNU General Public License for more details.
//
//        You should have received a copy of the GNU General Public License
//        along with this program.  If not, see <http://www.gnu.org/licenses/>.

// Establish connection
$host = $OMEGA["HOST"];
$user = $OMEGA["USER"];
$pass = $OMEGA["PASS"];
$conn = @mssql_connect($host, $user, $pass);
if (!$conn)
    return error("ERROR: %s", @mssql_get_last_message());


// Select database (if any)
if (isset($OMEGA["BASE"]))
{
    $select = @mssql_select_db($OMEGA['BASE'], $conn);
    if (!$select)
        return error("ERROR: %s", @mssql_get_last_message());
}


// Send query
$query = mssql_query($OMEGA['QUERY'], $conn);
if (!$query)
    return error("ERROR: %s", @mssql_get_last_message());


// Query type: GET (information gathering)
$rows = @mssql_num_rows($query);
if (is_int($rows))
{
    $result = array();
    $obj = mssql_fetch_array($query, MSSQL_ASSOC);
    $result[] = array_keys($obj);
    $result[] = array_values($obj);
    while ($line = mssql_fetch_array($query, MSSQL_ASSOC))
        $result[] = array_values($line);
    return array('GET', $rows, $result);
}


// Query type: SET (write into the database)
$rows = @mssql_rows_affected();
if (is_int($rows))
    return array('SET', $rows);

return error("ERROR: %s", @mssql_get_last_message());

?>

File: src/api/server/payload.py
#!/usr/bin/env python3

#            ---------------------------------------------------
#                              Omega Framework                                
#            ---------------------------------------------------
#                  Copyright (C) <2020>  <Entynetproject>       
#
#        This program is free software: you can redistribute it and/or modify
#        it under the terms of the GNU General Public License as published by
#        the Free Software Foundation, either version 3 of the License, or
#        any later version.
#
#        This program is distributed in the hope that it will be useful,
#        but WITHOUT ANY WARRANTY; without even the implied warranty of
#        MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
#        GNU General Public License for more details.
#
#        You should have received a copy of the GNU General Public License
#        along with this program.  If not, see <http://www.gnu.org/licenses/>.

import metadict
from core import session, tunnel, plugins
from datatypes import Path


class Payload(metadict.MetaDict):
    # the omega env vars to auto add
    # to $OMEGA array on php side
    _unherited_env_vars = ["PATH_SEP"]

    _php_vars_template = 'global $OMEGA;$OMEGA=%s;\n'

    def __init__(self, filename, **kwargs):
        self.response = None
        for key in self._unherited_env_vars:
            self[key] = session.Env[key]
        for key, value in kwargs.items():
            self[key] = value
        plugin_path = plugins.current_plugin.path
        self.payload = Path(plugin_path, filename, mode='fr').phpcode()

    def send(self, **kwargs):
        var_list = dict(self)
        for key, value in kwargs.items():
            var_list[key] = value
        php_vars = self._php_vars_template % tunnel.payload.py2php(var_list)

        result = tunnel.send(php_vars + self.payload)
        if result.response_error:
            raise PayloadError(result.response_error)
        return result.response


class PayloadError(Exception):
    """Exception raised when a send payload returned an __ERROR__ obj"""
